### ISSUE TITLE
Thumbnails for flash posts

### ISSUE THREAD

**Issue by henke37** [danbooru/danbooru#2002]
It would be nice if flash posts could have a thumbnail. The flash player is supposed to be scriptable by webcrawlers, including screenshots and interaction.

**Comment by ToksT** [comment]
I implemented it in a branch. It uses gnash to take a temporary full size screenshot of the *second* frame of the flash, then uses Danbooru's normal resizing code to scale it down. The reason for temporary screenshot is that if you use gnash to directly scale it down while taking the screenshot the quality is poor due to gnash's scaling algorithm. The reason for taking the second frame is that several flashes have a blank screen for the first frame. A couple still have a blank screen even on frame 2+, not sure if there's anything we can do for these.

@r888888888 To get it running you need to install gnash on the server. Also double check that it runs with just the command `gnash` (may or may not need to manually set that up, I had to).

Maybe put on testbooru? None of the flashes I've tried have had errors but I only tested 12 random ones from Danbooru.

**Comment by henke37** [comment]
IIRC Gnash is limited to avm0 files. That means no as3 files, which have only been growing in popularity since its introduction.

**Comment by Type-kun** [comment]
Hm. I have to say that if flash posts will have normal thumbnails, it would be nice to somehow visually distinguish them from regular animation because they might contain sound, be games etc. Perhaps "F" upper-left icon instead of triangle?

**Comment by ToksT** [comment]
> IIRC Gnash is limited to avm0 files. That means no as3 files, which have only been growing in popularity since its introduction.

Oh you're right, it seems the version 10 and higher swfs render either as a black screen, or they do render the animation but it's not entirely correct.

Is there any way to generate thumbnails for these swfs?

If not we should be able to detect the version by looking at the 4th byte of the file, so if it's 10 or above we can conditionally disable creating a thumbnail. But it's not ideal having this feature not work for a number of newer posts.

> Hm. I have to say that if flash posts will have normal thumbnails, it would be nice to somehow visually distinguish them from regular animation because they might contain sound, be games etc. Perhaps "F" upper-left icon instead of triangle?

Yeah, that's a good idea. Alternatively a speaker icon to indicate sound.

**Comment by henke37** [comment]
A speaker icon would be a bad idea, at least if no actual check for such content is done. And it's probably the halting problem to do the check flawlessly. But as long as there is no runtime decryption shenanigans, it's easy to check for the use of the sound tags and sound classes.

**Comment by Type-kun** [comment]
Use swftools perhaps? http://www.swftools.org/ , specifically http://wiki.swftools.org/wiki/Swfextract - I'm not sure how well this would work, but it allows image and frame extraction, perhaps these could be post-processed to get previews?

**Comment by Type-kun** [comment]
Oh, and there also seems to be an option to use `ffmpegthumbnailer`, with the only exception that it can't handle compressed SWFs; but there are tools to uncompress SWF files, it seems, including swfextract mentioned above. See http://stackoverflow.com/a/26404432/1375470 and http://askubuntu.com/a/22172

**Comment by ToksT** [comment]
> Use swftools perhaps? http://www.swftools.org/ , specifically http://wiki.swftools.org/wiki/Swfextract - I'm not sure how well this would work, but it allows image and frame extraction, perhaps these could be post-processed to get previews?

swfextract extracts files contained within the swf, like jpgs, pngs, mp3s and such. It will probably work for swfs that simply store images as animation frames, but it's not going to for games and vector animation and such.

**Comment by ToksT** [comment]
> Oh, and there also seems to be an option to use ffmpegthumbnailer, with the only exception that it can't handle compressed SWFs; but there are tools to uncompress SWF files, it seems, including swfextract mentioned above. See http://stackoverflow.com/a/26404432/1375470 and http://askubuntu.com/a/22172

Can't find windows binaries for ffmpegthumbnailer, but since I assume it just uses ffmpeg I tried that instead.

ffmpeg can render a thumbnail for swf files if you uncompress them first (e.g. `swfcombine -d compressed.swf -o uncompressed.swf`). But it seems ffmpeg has the same rendering problems as gnash with newer swf versions.

**Comment by henke37** [comment]
Yeah, I wouldn't expect as3 compatibility from code that doesn't even deal with compressed files.

**Comment by Type-kun** [comment]
Okay, it's a crazy idea, but it seems there's a possibility to run chromium in headless mode and take a screenshot. Sure seems like an overkill at the first glance, though.

**Comment by ddevault** [comment]
There's a new(ish) project called [Ruffle](https://github.com/ruffle-rs/ruffle) which can take a screenshot of the nth frame (5th seems to be a sweet spot for me). It's under active development and has reasonably broad support.

The catch is that it requires hardware accelleration and Vulkan. Works fine for me on AMDGPU w/mesa-vulkan-ati, but who knows what kind of GPU resources are available on donmai.us.
